### PR TITLE
[3D] Correctly compute near/far planes + culling when terrain offset is not null

### PR DESCRIPTION
**Pull request by ptitjano**
## Description

When the terrain offset is not null, the culling methods and near/far planes computations are wrong because they only rely on the bounding boxes which do not take into account the offset. 

This is fixed by adding a `terrainElevationOffset` method to the entities and adding it to the near/far plane computation and culling logic.

A new unit test for terrain offset is also added.

### Wrong behavior: current master

[offset_wrong_compress.webm](https://github.com/qgis/QGIS/assets/497207/d2d56186-6030-44a7-a897-9c8837cfd9f9)

### Fixed behavior: this PR

[offset_ok_compress.webm](https://github.com/qgis/QGIS/assets/497207/4edf90ab-4cfb-4341-b42e-04ac7e1c0e9f)

cc @benoitdm-oslandia @soaubier

### CODE REVIEW

File: src/3d/qgs3dutils.cpp [hunk 1]
```diff
@@ -558,14 +558,14 @@ static inline uint outcode( QVector4D v )
...
-bool Qgs3DUtils::isCullable( const QgsAABB &bbox, const QMatrix4x4 &viewProjectionMatrix )
+bool Qgs3DUtils::isCullable( const QgsAABB &bbox, const QMatrix4x4 &viewProjectionMatrix, float verticalOffset )
 {
   uint out = 0xff;
 
   for ( int i = 0; i < 8; ++i )
   {
     const QVector4D p( ( ( i >> 0 ) & 1 ) ? bbox.xMin : bbox.xMax,
-                       ( ( i >> 1 ) & 1 ) ? bbox.yMin : bbox.yMax,
+                       ( ( i >> 1 ) & 1 ) ? bbox.yMin + verticalOffset : bbox.yMax + verticalOffset,
```
Comment (nyalldawson) [inline src/3d/qgs3dutils.cpp:568]: This looks odd to me -- I think the passed bbox should just be shifted prior to passing to this function

Comment (ptitjano) [reply src/3d/qgs3dutils.cpp]: Done.

**Comment by uclaros** [comment]
This was on my radar for some time. Should we not just update the entities' bboxes instead so that `bbox()` returns their actual adjusted bbox?

**Comment by wonder-sk** [comment]
Thanks for looking into this issue - but indeed the correct solution would be what @uclaros suggests.

QgsChunkedEntity at this point has no ties to terrain whatsoever, and I think that should stay like that - we should not need to have a virtual function `terrainElevationOffset()` in there. So rather than doing any modifications in QgsChunkedEntity, we should be modifying terrain-related code to return correct bboxes (otherwise if bboxes are not correct, this kind of problem will appear elsewhere as well). I guess we'll need to modify places like:
- QgsTerrainGenerator::rootChunkBbox() - include terrain offset there
- QgsTerrainEntity - stop using mTerrainTransform and leave it for individual chunk entities to set up correct transform
- QgsDemTerrainTileLoader::createEntity() - apply vertical offset in its transform + include vertical offset when updating node's bbox with setExactBbox() call

I would also suggest to remove QgsTerrainEntity::terrainElevationOffset() that exists already, it is only used in QgsCameraController::setViewFromTop() and it is just a proxy for Qgs3DMapSettings::terrainElevationOffset() anyway...

(by the way, the QgsRuleBasedChunkedEntity::applyTerrainOffset() implementation would be best suited for a separate PR, as it is not directly related to the main problem this PR is trying to solve...)

**Comment by ptitjano** [comment]
> Thanks for looking into this issue - but indeed the correct solution would be what @uclaros suggests.
> 
> QgsChunkedEntity at this point has no ties to terrain whatsoever, and I think that should stay like that - we should not need to have a virtual function `terrainElevationOffset()` in there. So rather than doing any modifications in QgsChunkedEntity, we should be modifying terrain-related code to return correct bboxes (otherwise if bboxes are not correct, this kind of problem will appear elsewhere as well). I guess we'll need to modify places like:
> 
>     * QgsTerrainGenerator::rootChunkBbox() - include terrain offset there
> 
>     * QgsTerrainEntity - stop using mTerrainTransform and leave it for individual chunk entities to set up correct transform
> 
>     * QgsDemTerrainTileLoader::createEntity() - apply vertical offset in its transform + include vertical offset when updating node's bbox with setExactBbox() call
> 
> 
> I would also suggest to remove QgsTerrainEntity::terrainElevationOffset() that exists already, it is only used in QgsCameraController::setViewFromTop() and it is just a proxy for Qgs3DMapSettings::terrainElevationOffset() anyway...
> 
> (by the way, the QgsRuleBasedChunkedEntity::applyTerrainOffset() implementation would be best suited for a separate PR, as it is not directly related to the main problem this PR is trying to solve...)

Thanks for your comment. However, there is something I don't fully understand. Based on your suggestion, I understand that only the bboxes of the terrain need to be updated. However, the bounding boxes of `QgsVectorLayerChunkedEntity` and `QgsRuleBasedChunkedEntity` need to be upated because these entities are also translated if the terrain offset applies to them. That's the main reason for introducing the `QgsRuleBasedChunkedEntity::applyTerrainOffset()` commit and the introduction of `terrainElevationOffset()`

**Comment by ptitjano** [comment]
>     * QgsTerrainEntity - stop using mTerrainTransform and leave it for individual chunk entities to set up correct transform
 I also don't understand why you want to remove `mTerrainTransform` and replace with one by individual chunk entities. Could you please explain?

**Comment by ptitjano** [comment]
@wonder-sk gentle ping.

**Comment by benoitdm-oslandia** [comment]
Happy new year!

**Comment by wonder-sk** [comment]
> Based on your suggestion, I understand that only the bboxes of the terrain need to be updated. However, the bounding boxes of QgsVectorLayerChunkedEntity and QgsRuleBasedChunkedEntity need to be upated because these entities are also translated if the terrain offset applies to them. 

Yeah those will likely need updating as well to include the terrain offset in their bounding boxes - my comment meant to just add some pointers where I would be expecting modifications, rather than a full list, sorry for not being clear about that.

> I also don't understand why you want to remove mTerrainTransform and replace with one by individual chunk entities. Could you please explain?

IIRC the `mTerrainTransform` will also transform bounding boxes entity (if you have display of bounding boxes enabled) because they are parented to the terrain entity, so if the transform is limited to individual chunk entities, the bounding boxes entity should then display correctly.

**Pull request closed without merging**